src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -888,12 +983,48 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
 
                         if (status.ErrorCode == SecurityStatusPalErrorCode.ContextExpired)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         throw new IOException(SR.net_io_decrypt, SslStreamPal.GetException(status));
                     }
+
+                    if (decryptedCount > 0)
+                    {
+                        // This will either copy data from rented buffer or adjust final buffer as needed.
+                        // In both cases _decryptedBytesOffset and _decryptedBytesCount will be updated as needed.
+                        processedLength += CopyDecryptedData(buffer);
+                        if (_decryptedBytesCount > 0)
```

Comment:
I'm still confused by this test. 

I think we want to check if processedLength == buffer.Length, as we do above after we call CopyDecryptedData. Why aren't we doing that here?

Reply:
processedLength is cumulative e.g. it covers length of all fragments we processed so far. 
This may be personal but feel the check here basically shows that we have unconsumed decrypted data and that suggest we don't need to process more.